src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -95,32 +95,51 @@ public AnalyzerConfigOptionsResult GetOptionsForSourcePath(string sourcePath)
 
             var normalizedPath = PathUtilities.NormalizeWithForwardSlash(sourcePath);
 
-            // The editorconfig paths are sorted from shortest to longest, so matches
-            // are resolved from most nested to least nested, where last setting wins
+            var analyzerConfigsBuilder = ArrayBuilder<(int configIndex, AnalyzerConfig config)>.GetInstance(_analyzerConfigs.Length);
+
+            // The editorconfig paths are sorted from longest to shortest, so matches
+            // are resolved from least nested to most nested. This allows us to walk
+            // up the chain of editorconfig files until we find a root editorconfig
+            // or have processed them all.
             for (int analyzerConfigIndex = 0; analyzerConfigIndex < _analyzerConfigs.Length; analyzerConfigIndex++)
             {
                 var config = _analyzerConfigs[analyzerConfigIndex];
 
                 if (normalizedPath.StartsWith(config.NormalizedDirectory, StringComparison.Ordinal))
                 {
-                    int dirLength = config.NormalizedDirectory.Length;
-                    // Leave '/' if the normalized directory ends with a '/'. This can happen if
-                    // we're in a root directory (e.g. '/' or 'Z:/'). The section matching
-                    // always expects that the relative path start with a '/'. 
-                    if (config.NormalizedDirectory[dirLength - 1] == '/')
+                    analyzerConfigsBuilder.Add((analyzerConfigIndex, config));
+
+                    if (config.IsRoot)
                     {
-                        dirLength--;
+                        break;
                     }
-                    string relativePath = normalizedPath.Substring(dirLength);
+                }
+            }
+
+            // Reverse the list of applicable configs so that we walk from furthest to the
+            // the source file to closest allowing options to be overwritten as we go.
```

Comment:
Rather than overwriting, if we just update addOptions to not overwrite keys but preserve the "closest" value does that also work and then you can avoid having to reverse the order?

Reply:
(this work as it is, not sure it'll feel more natural or not)

Reply:
👍 Good suggestion. Updated.